temporal-serviceclient/src/main/java/io/temporal/serviceclient/CloudServiceStubsOptions.java

```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.serviceclient;
+
+import io.grpc.ManagedChannel;
+import java.util.Objects;
+
+/**
+ * Options for cloud service.
+ *
+ * <p>WARNING: The cloud service is currently experimental.
+ */
+public final class CloudServiceStubsOptions extends ServiceStubsOptions {
+  public static final String DEFAULT_CLOUD_TARGET = "saas-api.tmprl.cloud:443";
+
+  private static final CloudServiceStubsOptions DEFAULT_INSTANCE =
+      newBuilder().validateAndBuildWithDefaults();
+
+  /** Version header if any. */
+  private final String version;
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static Builder newBuilder(CloudServiceStubsOptions options) {
+    // We intentionally only accept our options and not the base class of
+    // options to ensure our defaults were originally applied at some point
+    // when the options class was first created.
+    return new Builder(options);
+  }
+
+  public static CloudServiceStubsOptions getDefaultInstance() {
+    return DEFAULT_INSTANCE;
+  }
+
+  private CloudServiceStubsOptions(ServiceStubsOptions serviceStubsOptions, String version) {
+    super(serviceStubsOptions);
+    this.version = version;
+  }
+
+  /**
+   * @return Returns the version used for the version header if any.
+   */
+  public String getVersion() {
+    return version;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    if (!super.equals(o)) return false;
+    CloudServiceStubsOptions that = (CloudServiceStubsOptions) o;
+    return Objects.equals(version, that.version);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(super.hashCode(), version);
+  }
+
+  /** Builder is the builder for ClientOptions. */
+  public static class Builder extends ServiceStubsOptions.Builder<Builder> {
+    private String version;
+
+    private Builder() {
+      // Set defaults only in this constructor
+      setTarget(DEFAULT_CLOUD_TARGET);
+      setEnableHttps(true);
+    }
+
+    private Builder(CloudServiceStubsOptions options) {
+      super(options);
+      this.version = options.version;
+    }
+
+    /** Set a cloud operation service version. This sets the version header for each call. */
+    public Builder setVersion(String version) {
+      this.version = version;
+      return this;
+    }
+
+    /** Default is {@link #DEFAULT_CLOUD_TARGET}. See inherited method for more details. */
+    @Override
+    public Builder setTarget(String target) {
+      return super.setTarget(target);
+    }
+
+    @Override
+    public Builder setChannel(ManagedChannel channel) {
+      // Unset our defaults
+      setEnableHttps(false);
```

Comment:
Why do we need this? `setChannel ` should be mutually exclusive with `setEnableHttps `

Reply:
There is this check in `ServiceStubsOptions.Builder.validateAndBuildWithDefaults`:

```
      if (this.enableHttps && this.channel != null) {
        throw new IllegalStateException(
            "Only one of the 'enableHttps' or 'channel' options can be set at a time");
      }
```

So if someone wants to set a channel, they don't know they have to go unset `enableHttps`. Earlier incarnations didn't do defaulting `enableHttps` to `true` eagerly this way, but they were all ugly.

Reply:
Probably not a big deal, but in principle I don't like overriding options user may have set. Why do you need to set the target and https in the builder constructor? couldn't you set it in `validateAndBuildWithDefaults` if a channel is not set?  Or just set them for `DEFAULT_INSTANCE`?

Reply:
> couldn't you set it in `validateAndBuildWithDefaults` if a channel is not set?

This is what I had originally but I struggled with the logic. How can you tell whether someone set `enableHttps` to know if it should be overridden, it's a primitive boolean. If you change it to `Boolean` in the builder to maintain a null/unset state that doesn't help if you are using the primitive in the actual options because most people use `build` and then we internally use the `newBuilder` override that accepts options to copy then validate and set defaults. Or do you want `Boolean` even in the options and just change the `getEnableHttps` to `return enableHttps != null && enableHttps`? I can do that if we want.

Reply:
Wouldn't you leave `enableHttps` false in the constructor so it is only true if a user enabled it.

Reply:
No, we want cloud client to default to HTTPS enabled. There's no reasonable situation where you'd use it without TLS (and even if there is, it's by no means a good default).

Reply:
I don't know if ew do. If someone is choosing to not use `newCloudServiceStubs` then I would not set `HTTPS` by default. It isn't consistent with our other services and if a user is not using  `newCloudServiceStubs` it's hard to say how they are connecting.

Reply:
They have to set other options almost always (e.g. API key). Arguably `newCloudServiceStubs` shouldn't even exist. Here's how the test creates a client and how users most likely will too (we hope the version requirement will be removed in the future):

```java
    CloudOperationsClient client =
        CloudOperationsClient.newInstance(
            CloudServiceStubs.newServiceStubs(
                CloudServiceStubsOptions.newBuilder()
                    .addApiKey(() -> apiKey)
                    .setVersion(apiVersion)
                    .build()));
```

Unlike connecting to workflow/operator service, here we know the https setting that will be used like 99% of the time.

Reply:
Only API key at this time (and maybe that will always be the case, unsure). Yes, we could _technically_ have overloads that accept params and use the builder for you (same for workflow service too where we know common things people need).

Reply:
My vote is `newCloudServiceStubs` has a required parameter for the `API` key and sets all the parameters and that would be the encouraged way to create the cloud service stub.

Reply:
We are not requiring API key in any cloud ops client in any SDK. That the cloud API currently requires an API key in its experimental form doesn't mean it always will. We are trying to have users set API keys the same way for clients since they are expected to be the same API keys. So if you use a certain way to create a workflow client in a language, you can apply that knowledge to creating a cloud client.

Reply:
>We are not requiring API key in any cloud ops client in any SDK

I just think it should be required for `newCloudServiceStubs ` not the `CloudServiceStubIOptions`. I would just like the expected case, connecting to our cloud with an API key, to be as easy as possible. If the requirements change we can change the SDK API to reflect that. I don't think it is a big deal, hence I approved.

Reply:
> I would just like the expected case, connecting to our cloud with an API key, to be as easy as possible.

This is going to start being the expected case with workflow service I think too (though there is self-hosted). I think we should consider making API keys as easy to use for everyone at the same time (even if it's used more often in one place than another).